### PR TITLE
Compile/Design-Time Checks for Variable Scopes Don't Work Properly with Qubit Allocation Blocks

### LINKED ISSUE #932: Compile/Design-Time Checks for Variable Scopes Don't Work Properly with Qubit Allocation Blocks
**Describe the bug**

When using the qubit block allocation syntax, the bound variable is not correctly scoped by the compile/design-time checks for valid variables.

**To Reproduce**

The following code incorrectly produces no errors until runtime:
```qsharp
namespace input {
    open Microsoft.Quantum.Measurement;

    @EntryPoint()
    operation Main() : Unit {
        use q = Qubit() {
            H(q);
            MResetZ(q);
        };
        H(q);
        MResetZ(q);
    }
}
```

**Expected behavior**

The last two statements using `q` should give errors about invalid variable name `q` because `q` is out of scope at that point. The code should be treated equivalently to:

```qsharp
namespace input {
    open Microsoft.Quantum.Measurement;

    @EntryPoint()
    operation Main() : Unit {
        {
            use q = Qubit();
            H(q);
            MResetZ(q);
        };
        H(q);
        MResetZ(q);
    }
}
```

### PR DESCRIPTION
**Pull request by swernli**
Fixes #932

### CODE REVIEW

**Review by ScottCarda-MS — APPROVED** [review]

File: compiler/qsc_frontend/src/resolve.rs [hunk 1]
```diff
@@ -667,10 +667,22 @@ impl AstVisitor<'_> for With<'_> {
             }
             ast::StmtKind::Qubit(_, pat, init, block) => {
                 ast_visit::walk_qubit_init(self, init);
-                // The binding is valid after end of the statement.
-                self.resolver.bind_pat(pat, stmt.span.hi);
                 if let Some(block) = block {
-                    self.visit_block(block);
+                    self.with_scope(block.span, ScopeKind::Block, |visitor| {
```
Comment (minestarks) [inline compiler/qsc_frontend/src/resolve.rs:671]: I think you may want to use the `with_pat` helper here. This mimics what we do for, say, for loops - this is kind of the same idea where the binding happens outside the block, but only applies to the block.

https://github.com/microsoft/qsharp/blob/b8a94e8d417dba5a8aef256103c4a14abd5937f5/compiler/qsc_frontend/src/resolve.rs#L701-L703

File: compiler/qsc_frontend/src/resolve.rs [hunk 2]
```diff
@@ -667,10 +667,22 @@ impl AstVisitor<'_> for With<'_> {
...
+                        // The binding is valid after end of the statement, but scoped to the current block.
+                        visitor.resolver.bind_pat(pat, stmt.span.hi);
```
Comment (minestarks) [inline compiler/qsc_frontend/src/resolve.rs:673]: `block.span.lo` would be the more correct offset to use here (but if you take my suggestion to use `with_pat`, you don't need to worry about it)

**Review by minestarks — APPROVED** [review]

File: compiler/qsc_frontend/src/resolve.rs (hunk 1 above)
Comment (swernli) [reply compiler/qsc_frontend/src/resolve.rs:671]: Oh, great suggestion! I missed that I could use an existing function for that, definitely a better way to go.

File: compiler/qsc_frontend/src/resolve.rs (hunk 2 above)
Comment (swernli) [reply compiler/qsc_frontend/src/resolve.rs:673]: I'll use your other suggestion.

**Pull request merged**